File: modules/echo-mesh/android/src/main/java/expo/modules/echomesh/PendingTable.kt
package expo.modules.echomesh

import java.util.concurrent.ConcurrentHashMap

/**
 * In-flight outbound messages awaiting ACKs. Keyed by wire messageId
 * (sender-local sequence). One pending entry per send.
 */
internal class PendingTable {
  data class Entry(
    val jsId: String,
    val groupId: String,
    val body: String,
    val kind: Byte,
    val frame: Frame,
    val expectedAcks: Int,
    val ackedBy: MutableSet<String> = mutableSetOf(),
    var attempt: Int = 1,
  )

  private val map = ConcurrentHashMap<Long, Entry>()

  fun put(messageId: Long, e: Entry) { map[messageId] = e }
  fun get(messageId: Long): Entry? = map[messageId]
  fun remove(messageId: Long): Entry? = map.remove(messageId)

  /**
   * Returns the entry and whether this ack newly transitions to "all
   * peers acked" (so the caller can emit final "delivered" once).
   */
  fun ack(messageId: Long, ackingSenderId: String): Entry? {
    val e = map[messageId] ?: return null
    e.ackedBy.add(ackingSenderId)
    return e
  }
}

File: modules/echo-mesh/ios/PendingTable.swift
import Foundation

/// In-flight outbound messages awaiting ACKs. Mirrors PendingTable.kt.
final class PendingTable {
  final class Entry {
    let jsId: String
    let groupId: String
    let body: String
    let kind: UInt8
    let frame: Frame
    let expectedAcks: Int
    var ackedBy: Set<String> = []
    var attempt: Int = 1
    init(jsId: String, groupId: String, body: String, kind: UInt8, frame: Frame, expectedAcks: Int) {
      self.jsId = jsId; self.groupId = groupId; self.body = body
      self.kind = kind; self.frame = frame; self.expectedAcks = expectedAcks
    }
  }

  private var map: [UInt32: Entry] = [:]
  private let lock = NSLock()

  func put(_ messageId: UInt32, _ e: Entry) {
    lock.lock(); defer { lock.unlock() }
    map[messageId] = e
  }

  func get(_ messageId: UInt32) -> Entry? {
    lock.lock(); defer { lock.unlock() }
    return map[messageId]
  }

  func remove(_ messageId: UInt32) -> Entry? {
    lock.lock(); defer { lock.unlock() }
    return map.removeValue(forKey: messageId)
  }

  func ack(_ messageId: UInt32, by senderId: String) -> Entry? {
    lock.lock(); defer { lock.unlock() }
    let e = map[messageId]
    e?.ackedBy.insert(senderId)
    return e
  }
}
